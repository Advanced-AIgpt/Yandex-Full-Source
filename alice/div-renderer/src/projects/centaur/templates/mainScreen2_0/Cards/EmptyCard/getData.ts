import { IEmptyCardProps } from './types';
import { ICardDataAdapter } from '../types';

export const getEmptyCardData: ICardDataAdapter<IEmptyCardProps> = function getEmptyCardData(card, requestState) {
    if (card && card.VacantCardData) {
        return {
            type: 'empty',
            requestState,
        };
    }
    return null;
};
