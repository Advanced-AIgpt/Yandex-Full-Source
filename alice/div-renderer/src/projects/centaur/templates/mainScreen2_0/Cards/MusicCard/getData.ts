import { IMusicCardProps } from './types';
import { colorDarkGrey } from '../../../../style/constants';
import { ICardDataAdapter } from '../types';

const playlists: readonly [string, string][] = [
    ['Плейлист дня', '#31AD54'],
    ['Мне нравится', '#E53866'],
    ['Тайник', '#186FD3'],
    ['Подкасты', '#036292'],
    ['Дежавю', '#A019D1'],
    ['Премьера', '#D1673E'],
    ['Плейлист с Алисой', '#2D0B87'],
    ['Мой 2021', '#C20146'],
    ['Плейлист семьи', '#FE8853'],
];

export const getMusicCardData: ICardDataAdapter<IMusicCardProps> = function getMusicCardData(card, requestState) {
    if (typeof card.MusicCardData !== 'undefined' && card.MusicCardData !== null) {
        const data = card.MusicCardData;

        const knownPlaylist = playlists.find(el => el[0] === data.Name);

        const color = knownPlaylist ? knownPlaylist[1] : colorDarkGrey;

        return {
            type: 'music',
            name: data.Name,
            modified: data.Modified?.value,
            color,
            cover: data.Cover,
            actions: (card.Action && [
                {
                    log_id: 'music_card_action',
                    url: card.Action,
                },
            ]) || undefined,
            preventDefaultLoader: true,
            requestState,
        };
    }
    return null;
};
