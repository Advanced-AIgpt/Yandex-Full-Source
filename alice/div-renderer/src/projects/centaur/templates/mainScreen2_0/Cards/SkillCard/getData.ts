import { ISkillCardProps } from './types';
import { ICardDataAdapter } from '../types';
import { NAlice } from '../../../../../../protos';
import { EnumLayer } from '../../../../actions/client';
type ITButton = NAlice.NData.TCentaurWidgetCardData.TExternalSkillCardData.TMainScreenData.ITButton;

interface INormalisedButton {
    Text: string;
    Payload?: string;
}

function isNormalButton(el: ITButton): el is INormalisedButton {
    return Boolean(el.Text);
}

export const getSkillCardData: ICardDataAdapter<ISkillCardProps> = function getSkillCardData(card, requestState) {
    if (typeof card.ExternalSkillCardData !== 'undefined' && card.ExternalSkillCardData !== null) {
        const data = card.ExternalSkillCardData;

        if (data.widgetGalleryData) {
            return {
                type: 'skill',
                requestState,
                skillName: data.skillInfo?.Name || ' ',
                skillImage: data.skillInfo?.Logo || ' ',
                skillId: data.skillInfo?.SkillId || ' ',
                title: ' ',
            };
        } else {
            return {
                type: 'skill',
                requestState,
                title: data.mainScreenData?.Title,
                comment: data.mainScreenData?.Text,
                image: data.mainScreenData?.ImageUrl,
                skillName: data.skillInfo?.Name || ' ',
                skillImage: data.skillInfo?.Logo || ' ',
                skillId: data.skillInfo?.SkillId || ' ',
                layer: EnumLayer.dialog,
                buttons: data.mainScreenData?.buttons
                    ?.filter(isNormalButton)
                    .map(el => ({
                        text: el.Text,
                        payload: el.Payload || undefined,
                    })),
            };
        }
    }
    return null;
};
