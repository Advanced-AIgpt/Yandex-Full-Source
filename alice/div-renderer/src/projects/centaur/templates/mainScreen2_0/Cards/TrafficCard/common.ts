import { ITrafficForecast } from './types';

export function sortTimeOfDay(originalArray: Readonly<ITrafficForecast[]>): ITrafficForecast[] {
    const forecastsFirst = [];
    const forecastsLast = [];

    for (const forecastElement of originalArray) {
        if (forecastElement.time === null || typeof forecastElement.time === 'undefined') {
            continue;
        }

        if (forecastsFirst.length === 0) {
            forecastsFirst.push(forecastElement);
        } else if (forecastsLast.length > 0) {
            forecastsLast.push(forecastElement);
        } else if (forecastElement.time - (forecastsFirst[forecastsFirst.length - 1].time as number) === 1) {
            forecastsFirst.push(forecastElement);
        } else {
            forecastsLast.push(forecastElement);
        }
    }

    return [...forecastsLast, ...forecastsFirst];
}
