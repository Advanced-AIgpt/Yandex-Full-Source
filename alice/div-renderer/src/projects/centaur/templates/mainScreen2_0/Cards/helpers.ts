export function getCardMainScreenId({ rowIndex, colIndex }: { rowIndex?: number; colIndex?: number }) {
    return `main_screen_card_${rowIndex || 0}_${colIndex || 0}`;
}
