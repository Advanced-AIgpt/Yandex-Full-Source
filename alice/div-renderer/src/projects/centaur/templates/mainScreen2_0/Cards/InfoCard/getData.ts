import { sample } from 'lodash';
import { IInfoCardProps } from './types';
import { DISCOVERY_CARDS } from '../../../mainScreen/discovery';
import { ICardDataAdapter } from '../types';
import { textAction } from '../../../../../../common/actions';

export const getInfoCardData: ICardDataAdapter<IInfoCardProps> = function getInfoCardData(card, requestState) {
    if (typeof card.InfoCardData !== 'undefined' && card.InfoCardData !== null) {
        const data = card.InfoCardData.Title === 'test' ? DISCOVERY_CARDS[0] : sample(DISCOVERY_CARDS) || DISCOVERY_CARDS[0];

        return {
            type: 'info',
            title: data.title,
            image_background: data.bg,
            description: data.description,
            requestState,
            actions: [
                {
                    log_id: 'discovery.mainscreen.execute',
                    url: textAction(data.description),
                },
            ],
        };
    }
    return null;
};
