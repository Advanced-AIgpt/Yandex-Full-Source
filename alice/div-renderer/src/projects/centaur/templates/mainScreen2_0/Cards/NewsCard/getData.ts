import { compact } from 'lodash';
import { INewsCardProps } from './types';
import { EnumLayer } from '../../../../actions/client';
import { getImageByTopic } from '../../../../helpers/newsHelper';
import { ICardDataAdapter } from '../types';

export const getNewsCardData: ICardDataAdapter<INewsCardProps> = function getNewsCardData(card, requestState) {
    if (typeof card.NewsCardData !== 'undefined' && card.NewsCardData !== null) {
        const data = card.NewsCardData;

        return {
            type: 'news',
            title: data.Title,
            content: data.Content,
            image: data.ImageUrl || getImageByTopic(data.Topic) || undefined,
            layer: EnumLayer.dialog,
            requestState,
            actions: compact([
                card.Action && {
                    log_id: 'news_card_action',
                    url: card.Action,
                },
            ]),
        };
    }
    return null;
};
