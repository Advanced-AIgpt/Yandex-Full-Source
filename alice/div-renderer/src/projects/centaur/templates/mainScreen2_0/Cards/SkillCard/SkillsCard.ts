import {
    ContainerBlock,
    Div,
    FixedSize,
    ImageBackground,
    ImageBlock,
    MatchParentSize,
    SolidBackground,
    TextBlock,
} from 'divcard2';
import { text28m, title56m } from '../../../../style/Text/Text';
import { backgroundToImageWithText, colorWhiteOpacity10, colorWhiteOpacity50 } from '../../../../style/constants';
import { ISkillCardProps } from './types';
import { logger } from '../../../../../../common/logger';
import BasicErrorCard from '../BasicErrorCard';
import { Avatar } from '../../../../../../common/helpers/avatar';
import { BasicCard, basicCardBackground } from '../BasicCard';
import { directivesAction } from '../../../../../../common/actions';
import { createSemanticFrameAction } from '../../../../../../common/actions/server';
import { getCardMainScreenId } from '../helpers';

export const DeviceName = 'SmartSpeaker';

function getSkillTitle({ skillName, skillImage }: {
    skillName: string,
    skillImage: string,
}) {
    return new ContainerBlock({
        orientation: 'horizontal',
        items: [
            new ImageBlock({
                image_url: skillImage,
                width: new FixedSize({ value: 44 }),
                height: new FixedSize({ value: 44 }),
                border: {
                    corner_radius: 22,
                },
                margins: {
                    right: 16,
                },
            }),
            new TextBlock({
                ...text28m,
                max_lines: 1,
                text_color: colorWhiteOpacity50,
                text: skillName,
            }),
        ],
    });
}

function SkillBigButton({ text, payload }: {
    text: string;
    payload?: string;
}, skillId: string, options?: Omit<ConstructorParameters<typeof TextBlock>[0], 'text'>) {
    return new ContainerBlock({
        items: [
            new TextBlock({
                ...text28m,
                text,
                background: [
                    new SolidBackground({
                        color: colorWhiteOpacity10,
                    }),
                ],
                width: new MatchParentSize({ weight: 1 }),
                height: new MatchParentSize({ weight: 1 }),
                text_alignment_horizontal: 'center',
                text_alignment_vertical: 'center',
                border: {
                    corner_radius: 24,
                },
                ...options,
            }),
        ],
        actions: [
            {
                log_id: 'skill_action',
                url: directivesAction(
                    createSemanticFrameAction(
                        {
                            external_skill_fixed_activate_semantic_frame: {
                                fixed_skill_id: {
                                    string_value: skillId,
                                },
                                activation_command: {
                                    string_value: text,
                                },
                                payload: {
                                    string_value: payload ? payload : '',
                                },
                                activation_source_type: {
                                    string_value: 'widget_gallery',
                                },
                            },
                        },
                        'Dialogovo',
                        'alice.external_skill_widget.button',
                    ),
                ),
            },
        ],
    });
}

export default function SkillsCard({
    image,
    actions,
    title,
    comment,
    longtap_actions,
    skillName,
    skillImage,
    skillId,
    buttons,
    colIndex,
    rowIndex,
}: ISkillCardProps): Div {
    if (!skillName) {
        logger.error(new Error('The news card should have a skillName, but some of this was not transmitted.'));

        return BasicErrorCard({
            colIndex,
            rowIndex,
            title: 'Ошибка',
            description: 'Ошибка сети. Не могу показать навык. Проверьте интернет.',
            actions,
            longtap_actions,
        });
    }

    let items: Div[] = [];

    if (buttons && buttons?.length > 1) {
        items = [
            new ContainerBlock({
                orientation: 'horizontal',
                height: new MatchParentSize({ weight: 1 }),
                margins: {
                    top: 45,
                },
                items: [
                    SkillBigButton(buttons[0], skillId, {
                        margins: {
                            right: 12,
                        },
                    }),
                    SkillBigButton(buttons[1], skillId, {
                        margins: {
                            left: 12,
                        },
                    }),
                ],
            }),
        ];
    } else if (buttons && buttons.length === 1) {
        items = [
            new TextBlock({
                ...text28m,
                text: title || comment || ' ',
                max_lines: 2,
                height: new MatchParentSize({ weight: 1 }),
                text_alignment_vertical: 'center',
            }),
            SkillBigButton(buttons[0], skillId, {
                height: new FixedSize({ value: 72 }),
            }),
        ];
    } else if (title && comment) {
        items = [
            new TextBlock({
                ...(title.length < 6 ? title56m : text28m),
                text: title,
                max_lines: 2,
                height: new MatchParentSize({ weight: 1 }),
                text_alignment_vertical: 'bottom',
            }),
            new TextBlock({
                ...text28m,
                text: comment,
                alpha: 0.5,
                max_lines: 2,
            }),
        ];
    } else if (title || comment) {
        const text = (title || comment) as string;

        items = [
            new TextBlock({
                ...(text.length < 6 ? title56m : text28m),
                text,
                max_lines: 3,
                height: new MatchParentSize({ weight: 1 }),
                text_alignment_vertical: 'bottom',
            }),
        ];
    } else {
        const text = 'Что-то пошло не так.\nНавык пока не работает:(';

        items = [
            new TextBlock({
                ...(text.length < 6 ? title56m : text28m),
                text,
                max_lines: 3,
                height: new MatchParentSize({ weight: 1 }),
                text_alignment_vertical: 'bottom',
                text_color: colorWhiteOpacity50,
            }),
        ];
    }

    const cardData: Parameters<typeof BasicCard>[0] = {
        id: getCardMainScreenId({ rowIndex, colIndex }),
        items: [
            getSkillTitle({ skillName, skillImage }),
            ...items,
        ],
        actions,
        longtap_actions,
    };

    if (image) {
        cardData.background = [
            ...basicCardBackground,
            new ImageBackground({
                image_url: Avatar.fromUrl(image)?.setTypeName('380x214', 'ynews').toString() ?? image,
                scale: 'fill',
                preload_required: 1,
                content_alignment_horizontal: 'center',
                content_alignment_vertical: 'center',
            }),
            ...backgroundToImageWithText,
        ];
    }

    return BasicCard(cardData);
}
