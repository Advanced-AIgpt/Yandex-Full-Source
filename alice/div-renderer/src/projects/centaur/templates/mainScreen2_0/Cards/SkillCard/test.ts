import SkillsCard from './SkillsCard';
import { AnonymizeDataForSnapshot } from '../../../../../../common/helpers/dev';
import { NAlice } from '../../../../../../protos';
import { getSkillCardData } from './getData';
import { logger } from '../../../../../../common/logger';
const TCardData = NAlice.NData.TCentaurMainScreenMyScreenData.TColumn.TCardData;

jest.mock('../../../../../../common/logger');

describe('Main screen empty card', () => {
    it('should get actual data', () => {
        expect(getSkillCardData(new TCardData({}))).toBeNull();

        expect(getSkillCardData(new TCardData({
            YouTubeCardData: {},
        }))).toBeNull();

        expect(getSkillCardData(new TCardData({
            YouTubeCardData: {},
            ExternalSkillCardData: {
            },
        }))).not.toBeNull();

        expect(getSkillCardData(new TCardData({
            ExternalSkillCardData: {
                skillInfo: {
                },
                widgetGalleryData: {},
            },
        }))).not.toBeNull();
    });

    it('should match snapshot', () => {
        expect(AnonymizeDataForSnapshot(SkillsCard({
            type: 'skill',
            title: 'Медуза',
            comment: 'Tesla начала устанавливать терминалы Starlink на зарядных станциях Supercharger',
            skillName: 'Навык',
            skillImage: '',
        }))).toMatchSnapshot();

        expect(logger.error).toHaveBeenCalledTimes(0);
    });
});
