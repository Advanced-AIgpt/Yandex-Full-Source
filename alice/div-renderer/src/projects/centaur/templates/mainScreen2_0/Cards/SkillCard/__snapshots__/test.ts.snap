// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main screen empty card should match snapshot 1`] = `
Object {
  "background": Array [
    Object {
      "color": "#1aeaf6ec",
      "type": "solid",
    },
  ],
  "border": Object {
    "corner_radius": 28,
  },
  "height": Object {
    "type": "match_parent",
    "weight": 1,
  },
  "id": "main_screen_card_0_0",
  "items": Array [
    Object {
      "items": Array [
        Object {
          "border": Object {
            "corner_radius": 22,
          },
          "height": Object {
            "type": "fixed",
            "value": 44,
          },
          "image_url": "",
          "margins": Object {
            "right": 16,
          },
          "type": "image",
          "width": Object {
            "type": "fixed",
            "value": 44,
          },
        },
        Object {
          "font_size": 28,
          "font_weight": "medium",
          "line_height": 36,
          "max_lines": 1,
          "text": "Навык",
          "text_color": "#80ffffff",
          "type": "text",
        },
      ],
      "orientation": "horizontal",
      "type": "container",
    },
    Object {
      "font_size": 28,
      "font_weight": "medium",
      "height": Object {
        "type": "match_parent",
        "weight": 1,
      },
      "line_height": 36,
      "max_lines": 2,
      "text": "Медуза",
      "text_alignment_vertical": "bottom",
      "text_color": "#fff",
      "type": "text",
    },
    Object {
      "alpha": 0.5,
      "font_size": 28,
      "font_weight": "medium",
      "line_height": 36,
      "max_lines": 2,
      "text": "Tesla начала устанавливать терминалы Starlink на зарядных станциях Supercharger",
      "text_color": "#fff",
      "type": "text",
    },
  ],
  "paddings": Object {
    "bottom": 24,
    "left": 24,
    "right": 24,
    "top": 24,
  },
  "type": "container",
  "width": Object {
    "type": "match_parent",
    "weight": 1,
  },
}
`;
